XLA optimized Implementation of StaticCache with `Tensor Indexing API`

huzama opened the pull request:
Use the `Index_copy` method to update the static cache inplace and avoid recompilation during each iteration in XLA

# What does this PR do?
The PR focuses on avoiding repeated recompilation during each iteration in XLA by performing in-place updates on the static cache. Specifically, it replaces direct tensor indexing assignments with `index_copy_` method calls. This technique ensures that cache updates are executed without triggering full tensor recompilations, improving runtime performance.

### Code Changes

**Original Code:**
```python
k_out[:, :, cache_position] = key_states
v_out[:, :, cache_position] = value_states
```

**Updated Code:**
```python
k_out.index_copy_(2, cache_position, key_states)
v_out.index_copy_(2, cache_position, value_states)
```




## Before submitting
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?

 @zucchini-nlp @gante 



Edit: 

- #### Discussion: [31126](https://github.com/huggingface/transformers/issues/31126)
- The code was further updated to create a new list instead of performing in-place updates to avoid recompilation in XLA.

huzama committed:
Use the `Index_copy` method to update static cache inplace and avoid recompilation during each iteration in XLA

huzama committed:
Use Index_copy method to update static cache inplace and avoid recompilation during each iteration in XLA

LysandreJik commented:
cc @ArthurZucker as well

ArthurZucker commented on src/transformers/cache_utils.py:
```diff
@@ -792,8 +791,25 @@ def update(
 
-        k_out[:, :, cache_position] = key_states
-        v_out[:, :, cache_position] = value_states
+        if is_torch_xla_available(): # If torch_xla is available, do out-of-place operation on KV_Cache and create a new list
+            k_out = k_out.index_copy(2, cache_position, key_states)
+            v_out = v_out.index_copy(2, cache_position, value_states)
+
+            updated_key_cache = [
+                k_out if i == layer_idx else self.key_cache[i] for i in range(len(self.key_cache))
+            ]
+
+            updated_value_cache = [
+                v_out if i == layer_idx else self.value_cache[i] for i in range(len(self.value_cache))
+            ]
+
+            self.key_cache = updated_key_cache
+            self.value_cache = updated_value_cache
+
+            return k_out, v_out
+
+        k_out.index_copy_(2, cache_position, key_states)
+        v_out.index_copy_(2, cache_position, value_states)
```
IMO this is typically what we want to avoid: adding very specific code path for a single support. 
However what we can potentially do is add a new cache class, for XLA static generation where you can do whatever you want!

huzama replied:
Thank you, @ArthurZucker, for pointing out the issue with my unwanted commit.

I'd like to mention that my commit still manages to perform in-place updates to the cache, which offers better performance compared to the previous implementation, but it's still relatively slow. If you have any suggestions for a more efficient way to handle out-of-place operations on `self.{}_cache`, I would greatly appreciate your input.

ArthurZucker replied:
If index_copy is faster for static cache in general then we should go for it! 
What I meant by unwanted is simply in the context of changing / adapting something general for a specific usecase! 

I don't know exactly how XLA handles this, but if we want fast XLA support out of the box then it makes sense to create another cache class optimised for XLA no? 

cc @tengomucho

tengomucho replied:
@huzama I don't know if you are using this in a GPU or TPU environment. If you think it's faster, maybe you can provide a small test that shows the improvement of your change? If you only see an improvement on the TPU, you can contribute on `optimum-tpu`, so we integrate it there directly without impacting more generic cache code.

ArthurZucker reviewed (commented):
FYI @gante and @zucchini-nlp

huzama committed:
Revert "Use Index_copy method to update static cache inplace and avoid recompilation during each iteration in XLA"

This reverts commit 1ad0a9af4206c32b03e1adcf23545e6a17d72db0.

huzama committed:
Revert back StaticCache Class to original code

huzama committed:
Create new StaticCacheXLA class for XLA support

huzama committed:
Update StaticCacheXLA with .index_copy for out-of-place Operation

huzama committed:
get_seq_length in StaticCacheXLA uses out-of-place index_select operation.
This is necessary to for XLA as tensors are not materilzed yet

huzama committed:
Warn user to use cache_position when calling the forward path

torch.arange(past_length) where past_length keeps changing causes recompilation in XLA

huzama committed:
Inherit StaticCacheXLA from StaticCache instead for compatibilty with isinstance(past_key_value, StaticCache)

huzama commented:
### Summary of Changes

@tengomucho, I conducted experiments with both `.index_copy` and Python slicing methods on the GPU and found no significant performance differences. Consequently, I have reverted the `StaticCache` code to its original form.

Based on @ArthurZucker's suggestions, I created a new class, `StaticCacheXLA`, with several key differences from the simple `StaticCache` class:

1. **Updating Cache with Out-of-Place Operations with `index_copy`**
   _XLA lazy tensors perform well with out-of-place operations:_
   https://github.com/huggingface/transformers/blob/71efd4c1f7a2f3411d19aff4a7f811f331c7330a/src/transformers/cache_utils.py#L1007-L1014

2. **Get `seq_len` Out-of-Place with `index_select`**
   https://github.com/huggingface/transformers/blob/71efd4c1f7a2f3411d19aff4a7f811f331c7330a/src/transformers/cache_utils.py#L1027-L1030

### Performance Improvements

#### Architecture Design

Using SPMD on LLAMA 3 following the guide from [PyTorch's high-performance LLAMA 2 blog](https://pytorch.org/blog/high-performance-llama-2/) and a simple generate function with greedy decoding, I achieved the following results:

#### Results

These experiments were run on a TPU v3-8. The input sequence length was 256, and the maximum number of new tokens was 512. I can upload numbers on TPU v4-128 later.

As a result of these changes and optimizations, the new `StaticCacheXLA` class achieved a generation rate of 6.35 iterations per second with TPU utilization at 31%. In comparison, the original `StaticCache` achieved 5.5 iterations per second with TPU utilization at 27%.

- **StaticCacheXLA**: 6.35 it/s, TPU utilization: 31%
- **StaticCache**: 5.5 it/s, TPU utilization: 27%

ArthurZucker reviewed (commented):
in general looks ok, we need to make sure generate can use it so update the mapping to detect if xla is used then use this class WDYT?

huzama commented:
@ArthurZucker Thank you for the feedback. If I understand correctly, you are suggesting that we should dynamically map to the appropriate cache implementation based on whether XLA is available. Would the following update reflect your suggestion?

```python
class StaticCacheDefault(Cache):
    """Default static cache implementation."""
    pass

class StaticCacheXLA(Cache):
    """XLA-optimized static cache implementation."""
    pass

# Determine which StaticCache implementation to use based on the availability of XLA.
StaticCache = StaticCacheXLA if is_torch_xla_available() else StaticCacheDefault
```

tengomucho commented on src/transformers/cache_utils.py:
```diff
@@ -935,3 +934,108 @@ def get_max_length(self) -> Optional[int]:
+        k_out = self.key_cache[layer_idx]
+        v_out = self.value_cache[layer_idx]
+
+        k_out = k_out.index_copy(2, cache_position, key_states)
```
I think you could use `index_copy_` to do in-place index copy

tengomucho reviewed (changes requested):
Before Merging this, I would suggest to do another change. I think you can remove all methods except `update` and`get_seq_length`, that are the only two functions that are different from the parent class.

Anyway this seems great, a quick test on a TPU showed me a speedup of 10% in inference on gemma-2b!

huzama committed:
Refactor StaticCacheXLA to use index_copy_ for in-place operation and remove unnecessory code

tengomucho reviewed (approved):
LGTM!

huzama committed:
Merge branch 'main' into cache_utils_xla

gante reviewed (commented):
The core of the PR LGTM, thank you for working on it! 💛 

Open questions before merging:
1. Utilization in `generate`: as we are adding more `Cache` classes, we are noticing that automagically initializing a `Cache` inside `generate` is becoming tricky. For my end, all that we need is to confirm that we can use this class as
```py
xla_cache = StaticCacheXLA(...)
generate_outputs = model.generate(**inputs, past_key_values=xla_cache)
```
2. Tests :D Can we add a test to confirm that the new class doesn't change the model outputs? We can also test that the API described in 1. works in that test (example of a similar test: https://github.com/huggingface/transformers/blob/43ee58588be4dc754c9f0dea874437fe7201bf00/tests/test_cache_utils.py#L171)

gante commented:
@huzama awesome! Can we add it as a test? 💛 After the test is added, I'm more than happy to approve the PR!

gante commented:
@tengomucho if all that's needed is to replace slicing by `.index_copy_`, I'm happy with changing `StaticCache` itself -- as long as we add a comment like "`k_out.index_copy_(2, cache_position, key_states)` is equivalent to `k_out[:, :, cache_position] = key_states`, but with better generalized support", for readability

If that's the case, no new tests would be needed 🤗 @huzama

huzama commented:
@gante I agree that combining both classes into a single implementation is a better approach. I'll add comments to explain the code and will push the changes with the merged implementation.

huzama committed:
Merge StaticCache and StaticCacheXLA class

gante reviewed (approved):
LGTM, thank you for iterating 🤗

gante commented:
@huzama you'll need to run `make fixup` on your `transformers` root dir and push the changes to make our CI happy 🤗

huzama committed:
Ran 'make fixup'

gante commented:
@huzama CI is still not fully happy -- the unrelated failure is probably solved by rebasing the PR and force-pushing it 😞

huzama committed:
Merge branch 'huggingface:main' into cache_utils_xla

tengomucho reviewed (approved):
LGTM!

ArthurZucker commented on src/transformers/cache_utils.py:
```diff
@@ -868,7 +872,17 @@ def get_seq_length(self, layer_idx: Optional[int] = 0) -> int:
         # limit the check to the first batch member and head dimension.
         # TODO: deprecate this function in favor of `cache_position`
-        return (self.key_cache[layer_idx][0, 0].any(dim=-1)).sum()
+        key_cache = self.key_cache[layer_idx]
+        device = key_cache.device
+
+        # index_select(dim, index) performs the same operation as item = tensor[..., index, ...]
+        # but it is used for better generality and flexibility.
+        # For more information, refer to: https://pytorch.org/cppdocs/notes/tensor_indexing.html
+
+        item = key_cache.index_select(0, torch.tensor(0, device=device))
+        head = item.index_select(1, torch.tensor(0, device=device))
+
+        return head.any(dim=-1).sum()
```
TBH this will be deprecated anyways!

ArthurZucker reviewed (approved):
LGTM, could you make sure you run the slow tests for Llama compile, gemma compile etc ?

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31129). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

tengomucho commented:
Hey @huzama are you planning to clean this up to get it merged? Otherwise let me know, I will be happy to take over to make `StaticCache` more efficient on XLA!

huzama commented:
@tengomucho Thank you for taking over. I have been running some other experiments, and I lost track of this pull request. I appreciate your help in making StaticCache more efficient on XLA!

tengomucho commented:
I opened another PR so I could push on the branch. Closing this!
